pkg/volume/flocker/plugin.go

```diff
@@ -0,0 +1,214 @@
+	// Flocker agent will take care of this, there is nothing we can do here
+	return nil, nil
+}
+
+type flockerBuilder struct {
+	*flocker
+	exe      exec.Interface
+	opts     volume.VolumeOptions
+	readOnly bool
+}
+
+func (b flockerBuilder) GetPath() string {
+	return b.flocker.path
+}
+
+func (b flockerBuilder) SetUp() error {
+	return b.SetUpAt(b.flocker.datasetName)
+}
+
+// newFlockerClient uses environment variables and pod attributes to return a
+// flocker client capable of talking with the Flocker control service.
+func (b flockerBuilder) newFlockerClient() (*flockerClient.Client, error) {
+	host := getenvOrFallback("FLOCKER_CONTROL_SERVICE_HOST", defaultHost)
+	portConfig := getenvOrFallback("FLOCKER_CONTROL_SERVICE_PORT", strconv.Itoa(defaultPort))
+	port, err := strconv.Atoi(portConfig)
+	if err != nil {
+		return nil, err
+	}
+	caCertPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CA_FILE", defaultCACertFile)
+	keyPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CLIENT_KEY_FILE", defaultClientKeyFile)
+	certPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CLIENT_CERT_FILE", defaultClientCertFile)
+
+	c, err := flockerClient.NewClient(host, port, b.flocker.pod.Status.HostIP, caCertPath, keyPath, certPath)
+	return c, err
+}
+
+func (b *flockerBuilder) getMetaDir() string {
+	return path.Join(b.plugin.host.GetPodPluginDir(b.flocker.pod.UID, util.EscapeQualifiedNameForDisk(flockerPluginName)), b.datasetName)
+}
+
+/*
+SetUpAt will setup a Flocker volume following this flow of calls to the Flocker
+control service:
+
+1. Get the dataset id for the given volume name/dir
+2. It should already be there, if it's not the user needs to manually create it
+3. Check the current Primary UUID
+4. If it doesn't match with the Primary UUID that we got on 2, then we will
+   need to update the Primary UUID for this volume.
+5. Wait until the Primary UUID was updated or timeout.
+*/
+func (b flockerBuilder) SetUpAt(dir string) error {
+	if volumeutil.IsReady(b.getMetaDir()) {
+		return nil
+	}
+
+	c, err := b.newFlockerClient()
+	if err != nil {
+		return err
+	}
+
+	datasetID, err := c.QueryDatasetIDFromName(dir)
+	if err != nil {
+		return err
+	}
+
+	s, err := c.GetDatasetState(datasetID)
+	if err != nil {
+		return fmt.Errorf("The volume '%s' is not available in Flocker. You need to create this manually with Flocker CLI before using it.", dir)
+	}
+
+	primaryUUID, err := c.LookupPrimaryUUID()
+	if err != nil {
+		return err
+	}
+
+	if s.Primary != primaryUUID {
+		// We need to update the primary and wait for it to be ready
+		err = c.UpdateDatasetPrimary(datasetID, primaryUUID)
+		if err != nil {
+			return err
+		}
+
+		timeoutChan := time.NewTimer(timeoutWaitingForVolume).C
+		tickChan := time.NewTicker(tickerWaitingForVolume).C
+
+		for {
+			if s, err = c.GetDatasetState(datasetID); err == nil && s.Primary == primaryUUID {
+				b.flocker.path = s.Path
```

Comment:
I'm sorry, I don't follow. Line 199 is `return err` for the time out. How/why would you want to go there?

I assume this (`primaryUUID` has been updated) is the terminating state, therefore you should break here which would break the loop and start executing at line 207 (`b.flocker.path = s.Path...`), which is what we want, right?